### DISCUSSION TITLE
removeExtension on document

### DISCUSSION THREAD

**Discussion by jo-chemla** [donmccurdy/glTF-Transform#1471] [Q&A] [answered]
Hi there, 

I'm trying to decompress ktx2 textures to png for a bunch of glb mesh tiles of a 3d-tiles tileset. I'm relying on the `ktx extract` CLI to do so and `3d-tiles-tools tilesetProcessor.processTileContentEntries`, more context and code sample [here](https://github.com/CesiumGS/3d-tiles-tools/issues/144). 

Thanks to `targetDocument.transform( customTransform({}), )`, I can 
 - convert all ktx textures to png, 
 - set their mimeType to `image/png`, 
 - and correctly reference these new png images in the output document. 

However, this resulting document still has the `KHR_texture_basisu` extension referenced as `extensionsUsed` and `extensionsRequired`. If I manually remove them from the glb->gltf json, blender_gltf_io plugin can import the resulting mesh with texture perfectly. 

I've tried multiple iterations of document root scene traversing all children, but logging `listExtensions()` on each of these returns only empty arrays. I therefore assume I have to manually remove the extension from the gltf document itselft. I can set the extension to null on the texture or document root via `texture.setExtension('KHR_texture_basisu', null)`, but could not find the equivalent to remove an extension on the document. 

The last resort I can think of is using `io.writeJSON(document, {}).then(jsonDoc` and remove extensions directly on the json object, before writing back to binary. I would love it if there are simpler alternatives. Thanks a lot for your help!

**Comment by jo-chemla** [comment, marked as answer]
Self replying here in case anyone stumbles on this discussion: [This bit](https://github.com/donmccurdy/glTF-Transform/blob/9e10d0049ae7f0a306e66c88826cec5458361d06/packages/view/examples/dropzone.ts#L95) let me find what I was missing, `extension.dispose();`

```js
        document.getRoot()
            .listExtensionsUsed()
            .find((e) => e.extensionName === 'KHR_texture_basisu')
            ?.dispose();
```

**Reply by donmccurdy** [reply]
Yes, this is what I'll usually do. Disposing the root-level extension will also remove any extension-specific properties attached elsewhere (though there are none in the case of KHR_texture_basisu).

Probably less intuitive to use [`createExtension`](https://gltf-transform.dev/modules/core/classes/Document#createExtension) for an existing extension, but this will have the same effect:

```javascript
import { KHRTextureBasisu } from '@gltf-transform/extensions';

document.createExtension(KHRTextureBasisu).dispose();
```

**Reply by jo-chemla** [reply]
Ah good to also know this `createExtension().dispose`, thanks as always!
